File: appcore/src/main/java/mobi/cangol/mobile/Task.java
package mobi.cangol.mobile;


import android.os.Handler;
import android.os.Message;

import java.lang.ref.WeakReference;

/**
 * Created by xuewu.wei on 2018/5/3.
 */
public abstract class Task<R> implements Runnable {
    private InnerHandler handler;

    public Task() {
        this.handler = new InnerHandler(this);
    }

    @Override
    public void run() {
        final R r = call();
        Message.obtain(handler, 1, r).sendToTarget();
    }

    public abstract R call();

    public abstract void result(R r);

    public static class InnerHandler extends Handler {
        private final WeakReference<Task> taskWeakReference;

        public InnerHandler(Task task) {
            taskWeakReference = new WeakReference<>(task);
        }

        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            if (msg.what == 1 && taskWeakReference.get() != null) {
                taskWeakReference.get().result(msg.obj);
                taskWeakReference.clear();
            }
        }
    }
}

File: appcore/src/main/java/mobi/cangol/mobile/Task.kt
package mobi.cangol.mobile


import android.os.Handler
import android.os.Message
import java.lang.ref.SoftReference

/**
 * Created by xuewu.wei on 2018/5/3.
 */
abstract class Task<R> : Runnable {
    private val handler: InnerHandler<R>

    init {
        this.handler = InnerHandler(this)
    }

    override fun run() {
        Message.obtain(handler, 1, call()).sendToTarget()
    }

    abstract fun call(): R

    abstract fun result(r: R)

    class InnerHandler<R>(task: Task<R>) : Handler() {
        private val taskReference: SoftReference<Task<R>> = SoftReference(task)

        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            if (msg.what == 1) {
                var task = taskReference.get()
                task?.result(msg.obj as R)
                taskReference.clear()
            }
        }
    }
}
